pypesto/ensemble/covariance_analysis.py

```diff
@@ -0,0 +1,318 @@
+
+    Parameters
+    ==========
+    matrix_to_analyze:
+        symmetric matrix (typically a covariance matrix) of parameters or
+        predictions
+
+    normalize:
+        flag indicating whether the returned Eigenvalues should be normalized
+        with respect to the largest Eigenvalue
+
+    lower_absolute_cutoff:
+        Consider only eigenvalues of the covariance matrix above this cutoff
+        (only applied when only_separable_directions is True)
+
+    lower_relative_cutoff:
+        Consider only eigenvalues of the covariance matrix above this cutoff,
+        when rescaled with the largest eigenvalue
+        (only applied when only_separable_directions is True)
+
+    inv_upper_absolute_cutoff:
+        Consider only low eigenvalues of the covariance matrix with inverses
+        above of this cutoff
+        (only applied when only_identifiable_directions is True)
+
+    inv_upper_relative_cutoff:
+        Consider only low eigenvalues of the covariance matrix when rescaled
+        with the largest eigenvalue with inverses above of this cutoff
+        (only applied when only_identifiable_directions is True)
+
+    only_identifiable_directions:
+        return only identifiable directions according to inv_upper_cutoff
+
+    only_separable_directions:
+        return only separable directions according to lower_cutoff
+
+    Returns
+    =======
+    eigen_values:
+        Eigenvalues of the covariance matrix
+
+    eigen_vectors:
+        eigenvectors of the covariance matrix
+    """
+
+    # get the eigenvalue decomposition
+    eigen_vals, eigen_vectors = np.linalg.eigh(matrix_to_analyze)
+
+    # get a normalized version
+    rel_eigen_vals = eigen_vals / np.max(eigen_vals)
+
+    # If no filtering is wanted, we can return
+    if not only_identifiable_directions and not only_separable_directions:
+        # apply normlization
+        if normalize:
+            eigen_vals = rel_eigen_vals
+        return eigen_vals, eigen_vectors
+
+    if only_identifiable_directions and only_identifiable_directions:
+        raise Exception('Asking for only identiafiable and only separable '
+                        'directions at the same time makes no sense. The '
+                        'applied filters are mutually exclusive.')
+
+    # Separable directions are wanted: an upper pass filtering is needed
+    if only_separable_directions:
+        if lower_absolute_cutoff is not None and \
+                lower_relative_cutoff is not None:
+            above_cutoff = np.array([
+                eigen_vals[i_eig] > lower_absolute_cutoff and
+                rel_eigen_vals[i_eig] > lower_relative_cutoff
+                for i_eig in range(len(eigen_vals))
+            ])
+        elif lower_absolute_cutoff is not None:
+            above_cutoff = eigen_vals > lower_absolute_cutoff
+        elif lower_relative_cutoff is not None:
+            above_cutoff = rel_eigen_vals > lower_relative_cutoff
```

Comment:
Could both relative and absolute cutoffs be useful together? Looks like this `if-elif` will only allow for one to be used?

Reply:
It's designed that both will be used, if both are given (the if clause), and that only one is used, if only one is given (the elif-clauses)